src/analysis/normed/ring/seminorm.lean

```diff
@@ -0,0 +1,196 @@
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_zero := λ f, f.map_zero' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`. -/
+instance : has_coe_to_fun (ring_seminorm R) (λ _, R → ℝ) := ⟨λ p, p.to_fun⟩
+
+@[simp] lemma to_fun_eq_coe (p : ring_seminorm R) : p.to_fun = p := rfl
+
+@[ext] lemma ext {p q : ring_seminorm R} (h : ∀ x, p x = q x) : p = q :=
+fun_like.ext p q h
+
+instance : has_zero (ring_seminorm R) :=
+⟨{ mul_le' :=  λ _ _, eq.ge (zero_mul _),
+  ..add_group_seminorm.has_zero.zero }⟩
+
+lemma eq_zero_iff {p : ring_seminorm R} : p = 0 ↔ ∀ x, p x = 0 := fun_like.ext_iff
+lemma ne_zero_iff {p : ring_seminorm R} : p ≠ 0 ↔ ∃ x, p x ≠ 0 := by simp [eq_zero_iff]
+
+instance : inhabited (ring_seminorm R) := ⟨0⟩
+
+variables (p : ring_seminorm R)
+
+protected lemma nonneg : 0 ≤ p x := p.nonneg' _
+@[simp] protected lemma map_zero : p 0 = 0 := p.map_zero'
+protected lemma add_le : p (x + y) ≤ p x + p y := p.add_le' _ _
+@[simp] protected lemma neg : p (- x) = p x := p.neg' _
+protected lemma mul_le : p (x * y) ≤ p x * p y := p.mul_le' _ _
+
+end non_unital_ring
+
+section ring
+
+variables [ring R] (p : ring_seminorm R)
+
+lemma pow_le : ∀ {n : ℕ}, 0 < n → p (x ^ n) ≤ (p x) ^ n
+| 1 h := by simp only [pow_one]
+| (n + 2) h :=
+by simpa only [pow_succ _ (n + 1)] using le_trans (p.mul_le x _)
+  (mul_le_mul_of_nonneg_left (pow_le n.succ_pos) (p.nonneg _))
+
+lemma seminorm_one_eq_one_iff_ne_zero (hp : p 1 ≤ 1) :
+  p 1 = 1 ↔ p ≠ 0 :=
+begin
+  refine ⟨λ h, ne_zero_iff.mpr ⟨1, by {rw h, exact one_ne_zero}⟩, λ h, _⟩,
+  obtain ⟨x, hx⟩ := ne_zero_iff.mp h,
+  by_cases hp0 : p 1 = 0,
+  { have hx' : p x ≤ 0,
+    { rw ← mul_one x,
+      apply le_trans (p.mul_le x 1) _,
+      rw [hp0, mul_zero], },
+    exact absurd (le_antisymm hx' (p.nonneg x) ) hx },
+  { have h1 : p 1 * 1 ≤ p 1 * p 1,
+    { conv_lhs { rw ← one_mul (1 : R) },
+      convert p.mul_le 1 1,
+      rw mul_one, },
+    rw mul_le_mul_left (lt_of_le_of_ne (p.nonneg _) (ne.symm hp0)) at h1,
+    exact le_antisymm hp h1, }
+end
+
+end ring
+
+end ring_seminorm
+
+/-- The norm of a `non_unital_semi_normed_ring` as a `ring_seminorm`. -/
+def norm_ring_seminorm (R : Type*) [non_unital_semi_normed_ring R] :
+  ring_seminorm R :=
+{ to_fun    := norm,
+  map_zero' := norm_zero,
+  nonneg'   := norm_nonneg,
+  add_le'   := norm_add_le,
+  neg'      := norm_neg,
+  mul_le'   := norm_mul_le }
+
+/-- A function `f : R → ℝ` is a norm on a (nonunital) ring if it is a seminorm and `f x = 0`
+  implies `x = 0`. -/
+structure ring_norm (R : Type*) [non_unital_ring R] extends ring_seminorm R :=
+(ne_zero : ∀ x, x ≠ 0 → 0 < to_fun x)
+
+attribute [nolint doc_blame] ring_norm.to_ring_seminorm
+
+namespace ring_norm
+
+variable [non_unital_ring R]
+
+instance zero_hom_class : zero_hom_class (ring_norm R) R ℝ :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_zero := λ f, f.map_zero' }
```

Comment:
This should become an instance `ring_norm_class.to_add_group_norm_class`.

Reply:
Yes, but in a future PR. I am not adding `ring_norm_class` in this PR.

Reply:
Why? That sounds like it should go together. At least upgrade that to `add_group_norm_class (ring_norm R) R`.